File: back-end/app/domain/services/attachment_service.py
"""
Servicio de dominio para Attachments.
Contiene la lógica de negocio para archivos adjuntos de CTFs.
"""

from typing import Dict, List, Optional, Tuple
from uuid import UUID
from dataclasses import dataclass

from ..entities.attachment import Attachment, AttachmentType
from ..entities.ctf import CTFCategory
from ..repositories.attachment_repo import AttachmentRepository


@dataclass
class AttachmentConfig:
    """Configuración de adjuntos por categoría."""
    allowed_types: List[AttachmentType]
    max_size: int  # bytes
    max_files: int
    allowed_extensions: List[str]
    allowed_mime_types: List[str]


class AttachmentService:
    """Servicio de dominio para lógica de adjuntos."""
    
    # Configuración por categoría de CTF
    CATEGORY_CONFIGS: Dict[str, AttachmentConfig] = {
        CTFCategory.WEB.value: AttachmentConfig(
            allowed_types=[AttachmentType.URL],
            max_size=0,
            max_files=3,
            allowed_extensions=[],
            allowed_mime_types=[],
        ),
        CTFCategory.CRYPTO.value: AttachmentConfig(
            allowed_types=[AttachmentType.FILE],
            max_size=5 * 1024 * 1024,  # 5MB
            max_files=3,
            allowed_extensions=[".txt", ".enc", ".bin", ".json", ".pem", ".key"],
            allowed_mime_types=["text/plain", "application/octet-stream", "application/json"],
        ),
        CTFCategory.STEGO.value: AttachmentConfig(
            allowed_types=[AttachmentType.FILE],
            max_size=50 * 1024 * 1024,  # 50MB
            max_files=5,
            allowed_extensions=[".png", ".jpg", ".jpeg", ".gif", ".bmp", ".wav", ".mp3", ".mp4"],
            allowed_mime_types=["image/*", "audio/*", "video/*"],
        ),
        CTFCategory.FORENSICS.value: AttachmentConfig(
            allowed_types=[AttachmentType.FILE],
            max_size=100 * 1024 * 1024,  # 100MB
            max_files=5,
            allowed_extensions=[".pcap", ".pcapng", ".raw", ".mem", ".img", ".dd", ".png", ".jpg", ".wav"],
            allowed_mime_types=["application/vnd.tcpdump.pcap", "application/octet-stream", "image/*", "audio/*"],
        ),
        CTFCategory.REVERSE.value: AttachmentConfig(
            allowed_types=[AttachmentType.FILE],
            max_size=50 * 1024 * 1024,  # 50MB
            max_files=2,
            allowed_extensions=[".exe", ".elf", ".bin", ".so", ".dll", ".apk", ".jar"],
            allowed_mime_types=["application/x-executable", "application/x-elf", "application/octet-stream"],
        ),
        CTFCategory.PWN.value: AttachmentConfig(
            allowed_types=[AttachmentType.FILE, AttachmentType.DOCKER],
            max_size=50 * 1024 * 1024,  # 50MB
            max_files=2,
            allowed_extensions=[".elf", ".bin", ".c", ".py"],
            allowed_mime_types=["application/x-executable", "application/x-elf", "application/octet-stream", "text/plain"],
        ),
        CTFCategory.MISC.value: AttachmentConfig(
            allowed_types=[AttachmentType.FILE, AttachmentType.URL],
            max_size=20 * 1024 * 1024,  # 20MB
            max_files=5,
            allowed_extensions=["*"],  # Cualquiera
            allowed_mime_types=["*/*"],
        ),
        CTFCategory.OSINT.value: AttachmentConfig(
            allowed_types=[AttachmentType.URL, AttachmentType.FILE],
            max_size=10 * 1024 * 1024,  # 10MB
            max_files=3,
            allowed_extensions=[".png", ".jpg", ".jpeg"],
            allowed_mime_types=["image/*"],
        ),
    }
    
    def __init__(self, attachment_repository: AttachmentRepository):
        self.attachment_repository = attachment_repository
    
    def get_config_for_category(self, category: str) -> Optional[AttachmentConfig]:
        """Obtiene la configuración para una categoría."""
        return self.CATEGORY_CONFIGS.get(category)
    
    def validate_file(
        self,
        filename: str,
        size: int,
        mime_type: str,
        category: str,
    ) -> Tuple[bool, Optional[str]]:
        """
        Valida un archivo según la configuración de la categoría.
        
        Returns:
            Tuple de (es_válido, mensaje_error)
        """
        config = self.get_config_for_category(category)
        if not config:
            return False, "Categoría inválida"
        
        # Verificar que la categoría permite archivos
        if AttachmentType.FILE not in config.allowed_types:
            return False, "Esta categoría no acepta archivos"
        
        # Validar tamaño
        if config.max_size > 0 and size > config.max_size:
            max_mb = config.max_size / (1024 * 1024)
            return False, f"Archivo demasiado grande. Máximo: {max_mb:.1f}MB"
        
        # Validar extensión
        if "*" not in config.allowed_extensions:
            ext = self._get_extension(filename)
            if ext.lower() not in [e.lower() for e in config.allowed_extensions]:
                return False, f"Extensión no permitida. Permitidas: {', '.join(config.allowed_extensions)}"
        
        # Validar MIME type
        if "*/*" not in config.allowed_mime_types:
            if not self._is_mime_type_allowed(mime_type, config.allowed_mime_types):
                return False, f"Tipo de archivo no permitido: {mime_type}"
        
        return True, None
    
    def validate_url(self, url: str, category: str) -> Tuple[bool, Optional[str]]:
        """
        Valida una URL según la configuración de la categoría.
        
        Returns:
            Tuple de (es_válida, mensaje_error)
        """
        config = self.get_config_for_category(category)
        if not config:
            return False, "Categoría inválida"
        
        # Verificar que la categoría permite URLs
        if AttachmentType.URL not in config.allowed_types:
            return False, "Esta categoría no acepta URLs"
        
        # Validar formato de URL
        if not url or not (url.startswith("http://") or url.startswith("https://")):
            return False, "URL inválida. Debe comenzar con http:// o https://"
        
        return True, None
    
    def get_ctf_attachments(self, ctf_id: UUID) -> List[Attachment]:
        """Obtiene todos los adjuntos de un CTF."""
        return self.attachment_repository.get_by_ctf_id(ctf_id)
    
    def _get_extension(self, filename: str) -> str:
        """Obtiene la extensión de un archivo."""
        if "." not in filename:
            return ""
        return "." + filename.rsplit(".", 1)[-1]
    
    def _is_mime_type_allowed(self, mime_type: str, allowed: List[str]) -> bool:
        """Verifica si un MIME type está permitido."""
        for allowed_type in allowed:
            if allowed_type.endswith("/*"):
                # Wildcard: image/* permite image/png, image/jpeg, etc.
                prefix = allowed_type[:-1]
                if mime_type.startswith(prefix):
                    return True
            elif allowed_type == mime_type:
                return True
        return False
    
    def requires_files(self, category: str) -> bool:
        """Verifica si una categoría requiere archivos."""
        config = self.get_config_for_category(category)
        return config is not None and AttachmentType.FILE in config.allowed_types
    
    def requires_url(self, category: str) -> bool:
        """Verifica si una categoría requiere URLs."""
        config = self.get_config_for_category(category)
        if not config:
            return False
        # Requiere URL si es el único tipo permitido
        return config.allowed_types == [AttachmentType.URL]
    
    def supports_docker(self, category: str) -> bool:
        """Verifica si una categoría soporta Docker."""
        config = self.get_config_for_category(category)
        return config is not None and AttachmentType.DOCKER in config.allowed_types

File: front-end/portafolio/src/app/core/services/attachment.service.ts
/**
 * Attachment Service
 * Servicio para gestión de adjuntos (archivos y URLs) en retos CTF
 */

import { Injectable, inject } from '@angular/core';
import {
  CTFCategory,
  CTFAttachment,
  AttachmentType,
  CategoryAttachmentConfig,
  ValidationResult,
  CATEGORY_ATTACHMENT_CONFIG
} from '@core/models/ctf.model';
import { FileValidators } from '@core/validators/file.validators';
import { ApiService } from './api.service';
import { firstValueFrom } from 'rxjs'; // For toPromise replacement if strictly angular 16+ (using toPromise for now as it's common)

export interface FileUploadResult {
  success: boolean;
  attachment?: CTFAttachment;
  error?: string;
}

@Injectable({
  providedIn: 'root'
})
export class AttachmentService {

  constructor(private api: ApiService) { }

  // ============================================
  // CONFIGURACIÓN POR CATEGORÍA
  // ============================================

  /**
   * Obtiene la configuración de adjuntos para una categoría
   */
  getConfigForCategory(category: CTFCategory): CategoryAttachmentConfig {
    return CATEGORY_ATTACHMENT_CONFIG[category];
  }

  /**
   * Verifica si una categoría requiere archivos
   */
  requiresFiles(category: CTFCategory): boolean {
    const config = this.getConfigForCategory(category);
    return config?.requiredTypes.includes('file') ?? false;
  }

  /**
   * Verifica si una categoría requiere URLs
   */
  requiresUrl(category: CTFCategory): boolean {
    const config = this.getConfigForCategory(category);
    return config?.requiredTypes.includes('url') ?? false;
  }

  /**
   * Verifica si una categoría soporta Docker
   */
  supportsDocker(category: CTFCategory): boolean {
    const config = this.getConfigForCategory(category);
    return config?.requiredTypes.includes('docker') ?? false;
  }

  /**
   * Obtiene los tipos de adjuntos permitidos para una categoría
   */
  getAllowedTypes(category: CTFCategory): AttachmentType[] {
    const config = this.getConfigForCategory(category);
    return config?.requiredTypes ?? [];
  }

  // ============================================
  // VALIDACIÓN DE ARCHIVOS
  // ============================================

  /**
   * Valida un archivo contra la configuración de categoría
   */
  validateFile(file: File, category: CTFCategory): ValidationResult {
    const config = this.getConfigForCategory(category);
    const errors: string[] = [];

    if (!config) {
      return { valid: false, errors: ['Categoría no válida'] };
    }

    // No permite archivos
    if (!config.requiredTypes.includes('file')) {
      return { valid: false, errors: ['Esta categoría no acepta archivos'] };
    }

    // Validar tamaño
    if (config.maxFileSize > 0 && file.size > config.maxFileSize) {
      errors.push(`Archivo demasiado grande. Máximo: ${FileValidators.formatSize(config.maxFileSize)}`);
    }

    // Validar MIME type
    if (config.allowedMimeTypes.length > 0 && !config.allowedMimeTypes.includes('*/*')) {
      if (!FileValidators.isMimeTypeAllowed(file.type, config.allowedMimeTypes)) {
        errors.push(`Tipo de archivo no permitido: ${file.type || 'desconocido'}`);
      }
    }

    // Validar extensión
    const ext = FileValidators.getExtension(file.name);
    if (config.allowedExtensions.length > 0 && !config.allowedExtensions.includes('*')) {
      if (!FileValidators.isExtensionAllowed(ext, config.allowedExtensions)) {
        errors.push(`Extensión .${ext} no permitida. Usa: ${config.allowedExtensions.join(', ')}`);
      }
    }

    return { valid: errors.length === 0, errors };
  }

  /**
   * Valida múltiples archivos
   */
  validateFiles(files: File[], category: CTFCategory): ValidationResult {
    const config = this.getConfigForCategory(category);
    const errors: string[] = [];

    if (!config) {
      return { valid: false, errors: ['Categoría no válida'] };
    }

    // Validar cantidad máxima
    if (config.maxFiles > 0 && files.length > config.maxFiles) {
      errors.push(`Máximo ${config.maxFiles} archivos permitidos`);
    }

    // Validar cada archivo
    files.forEach((file, index) => {
      const result = this.validateFile(file, category);
      if (!result.valid) {
        errors.push(`Archivo ${index + 1} (${file.name}): ${result.errors.join(', ')}`);
      }
    });

    return { valid: errors.length === 0, errors };
  }

  // ============================================
  // VALIDACIÓN DE URLs
  // ============================================

  /**
   * Valida una URL contra la configuración de categoría
   */
  validateUrl(url: string, category: CTFCategory): ValidationResult {
    const config = this.getConfigForCategory(category);
    const errors: string[] = [];

    if (!config) {
      return { valid: false, errors: ['Categoría no válida'] };
    }

    // Validar formato URL
    try {
      new URL(url);
    } catch {
      return { valid: false, errors: ['URL inválida'] };
    }

    // Validar patrones si existen
    if (config.urlPatterns && config.urlPatterns.length > 0) {
      const matchesPattern = config.urlPatterns.some(p => p.test(url));
      if (!matchesPattern) {
        errors.push('La URL no cumple con el formato requerido');
      }
    }

    // Validar protocolo (debe ser http o https)
    const parsed = new URL(url);
    if (!['http:', 'https:'].includes(parsed.protocol)) {
      errors.push('Solo se permiten URLs HTTP o HTTPS');
    }

    return { valid: errors.length === 0, errors };
  }

  // ============================================
  // CREACIÓN DE ADJUNTOS
  // ============================================

  /**
   * Crea un objeto CTFAttachment desde un File
   */
  createFileAttachment(file: File, uploadedUrl: string): CTFAttachment {
    return {
      id: this.generateId(),
      name: file.name,
      type: 'file',
      url: uploadedUrl,
      size: file.size,
      mimeType: file.type,
      uploadedAt: new Date()
    };
  }

  /**
   * Crea un objeto CTFAttachment desde una URL
   */
  createUrlAttachment(url: string, name?: string): CTFAttachment {
    const parsed = new URL(url);
    return {
      id: this.generateId(),
      name: name || parsed.hostname,
      type: 'url',
      url: url,
      uploadedAt: new Date()
    };
  }

  /**
   * Crea un objeto CTFAttachment para Docker
   */
  createDockerAttachment(imageName: string): CTFAttachment {
    return {
      id: this.generateId(),
      name: imageName,
      type: 'docker',
      url: `docker://${imageName}`,
      uploadedAt: new Date()
    };
  }

  // ============================================
  // UPLOAD (simulado - integrar con backend real)
  // ============================================

  /**
   * Sube un archivo al servidor (simulado)
   * En producción: integrar con tu backend/storage
   */
  /**
   * Sube un archivo al servidor
   */
  async uploadFile(file: File, category: CTFCategory, ctfId?: string): Promise<FileUploadResult> {
    // Validar primero
    const validation = this.validateFile(file, category);
    if (!validation.valid) {
      return { success: false, error: validation.errors.join(', ') };
    }

    try {
      const formData = new FormData();
      formData.append('file', file);
      if (ctfId) {
        formData.append('ctf_id', ctfId);
      }

      const response = await firstValueFrom(this.api.upload<any>('/attachments/upload', formData));

      return {
        success: true,
        attachment: this.mapApiToAttachment(response)
      };
    } catch (error: any) {
      console.error('Error uploading file:', error);
      return {
        success: false,
        error: error.message || 'Error al subir archivo'
      };
    }
  }

  private mapApiToAttachment(apiResponse: any): CTFAttachment {
    return {
      id: apiResponse.id,
      name: apiResponse.filename,
      type: 'file',
      url: apiResponse.url,
      size: apiResponse.size,
      mimeType: apiResponse.mime_type,
      uploadedAt: new Date()
    };
  }

  /**
   * Sube múltiples archivos
   */
  async uploadFiles(files: File[], category: CTFCategory): Promise<FileUploadResult[]> {
    const results: FileUploadResult[] = [];

    for (const file of files) {
      const result = await this.uploadFile(file, category);
      results.push(result);
    }

    return results;
  }

  // ============================================
  // UTILIDADES
  // ============================================

  /**
   * Genera un ID único
   */
  private generateId(): string {
    return `att_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Formatea tamaño de archivo
   */
  formatFileSize(bytes: number): string {
    return FileValidators.formatSize(bytes);
  }

  /**
   * Obtiene las extensiones permitidas como string
   */
  getAllowedExtensionsString(category: CTFCategory): string {
    const config = this.getConfigForCategory(category);
    if (!config || config.allowedExtensions.includes('*')) {
      return 'Cualquier extensión';
    }
    return config.allowedExtensions.join(', ');
  }

  /**
   * Obtiene el tamaño máximo formateado
   */
  getMaxFileSizeFormatted(category: CTFCategory): string {
    const config = this.getConfigForCategory(category);
    if (!config || config.maxFileSize === 0) {
      return 'Sin límite';
    }
    return FileValidators.formatSize(config.maxFileSize);
  }

  /**
   * Obtiene el accept string para input file
   */
  getAcceptString(category: CTFCategory): string {
    const config = this.getConfigForCategory(category);
    if (!config || config.allowedMimeTypes.includes('*/*')) {
      return '*/*';
    }

    // Combinar MIME types y extensiones
    const mimes = config.allowedMimeTypes.join(',');
    const exts = config.allowedExtensions
      .filter(e => e !== '*')
      .map(e => e.startsWith('.') ? e : `.${e}`)
      .join(',');

    return [mimes, exts].filter(Boolean).join(',');
  }
}
